merkletree.go

```diff
@@ -0,0 +1,274 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+	"github.com/coniks-sys/libmerkleprefixtree-go/internal"
+)
+
+var (
+	ErrInvalidTree = errors.New("[merkletree] invalid tree")
+)
+
+const (
+	EmptyBranchIdentifier = 'E'
+	LeafIdentifier        = 'L'
+)
+
+type MerkleTree struct {
+	treeNonce []byte
+	salt      []byte
+	privKey   []byte
+	pubKey    []byte
+	policies  Policies
+	root      *interiorNode
+}
+
+func InitMerkleTree(policies Policies, treeNonce, salt, pubKey, privKey []byte) *MerkleTree {
+	leftBranch := &emptyNode{
+		node: node{
+			level: 1,
+		},
+	}
+
+	rightBranch := &emptyNode{
+		node: node{
+			level: 1,
+		},
+	}
+	root := &interiorNode{
+		node: node{
+			parent: nil,
+			level:  0,
+		},
+		leftChild:  leftBranch,
+		rightChild: rightBranch,
+		leftHash:   nil,
+		rightHash:  nil,
+	}
+	leftBranch.parent = root
+	rightBranch.parent = root
+
+	m := &MerkleTree{
+		treeNonce: treeNonce,
+		salt:      salt,
+		privKey:   privKey,
+		pubKey:    pubKey,
+		policies:  policies,
+		root:      root,
+	}
+	return m
+}
+
+func (h *History) LookUp(key string) (MerkleNode, []LookUpProofNode, error) {
+	str := h.currentSTR
+	return lookUp(key, str)
+}
+
+func (h *History) LookUpInEpoch(key string, ep int64) (MerkleNode, []LookUpProofNode, error) {
+	str := h.GetSTR(ep)
+	return lookUp(key, str)
+}
+
+func lookUp(key string, str *SignedTreeRoot) (MerkleNode, []LookUpProofNode, error) {
+	lookupIndex := computePrivateIndex(key)
+	position := 0
+	var nodePointer interface{}
+	nodePointer = str.treeRoot
+	var proof []LookUpProofNode
+
+	for {
+		if _, ok := nodePointer.(*userLeafNode); ok {
+			// reached to a leaf node
+			break
+		}
+		if _, ok := nodePointer.(*emptyNode); ok {
+			// reached to an empty branch
+			break
+		}
+		proof = append(proof, nodePointer.(*interiorNode))
+		direction := util.GetNthBit(lookupIndex, position)
+		if direction {
+			nodePointer = nodePointer.(*interiorNode).rightChild
+		} else {
+			nodePointer = nodePointer.(*interiorNode).leftChild
+		}
+		position++
+	}
+
+	if nodePointer == nil {
+		return nil, nil, ErrInvalidTree
+	}
+	switch nodePointer.(type) {
+	case *userLeafNode:
+		proof = append(proof, nodePointer.(*userLeafNode))
+		if nodePointer.(*userLeafNode).key == key {
+			return nodePointer.(*userLeafNode), proof, nil
+		}
+		// reached a different leaf with a matching prefix
+		// return nil and a auth path including the leaf node
+		return nil, proof, nil
+	case *emptyNode:
+		proof = append(proof, nodePointer.(*emptyNode))
```

Comment:
> But we don't need to return the path, right. Since we're returning a node, you can walk it back from the parent property.
Yes. But the `MerkleNode` exposes some methods that are not expected in auth path. The `MerkleNode` is currently used to read the value of the node return by `Get` whereas the `LookUpProofNode` (I renamed to `ProofNode` in the latest commit) is expected to return some fields (e.g. the hash of the node currently) to preserve the privacy.